llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp

```diff
@@ -3847,15 +3847,36 @@ void DAGTypeLegalizer::ExpandIntRes_CTLZ(SDNode *N,
   Hi = DAG.getConstant(0, dl, NVT);
 }
 
-void DAGTypeLegalizer::ExpandIntRes_CTPOP(SDNode *N,
-                                          SDValue &Lo, SDValue &Hi) {
-  SDLoc dl(N);
+void DAGTypeLegalizer::ExpandIntRes_CTPOP(SDNode *N, SDValue &Lo, SDValue &Hi) {
+  SDValue Op = N->getOperand(0);
+  EVT VT = N->getValueType(0);
+  SDLoc DL(N);
+
+  // If the narrow CTPOP is not supported by the target, try to convert it
```

Comment:
Do we really have to handle the libcall logic during type legalization? Can this just adjust the type, and leave the libcall for the libcall action

Reply:
If we expand it later, we will end up with two libcalls for an illegal type instead of one (best case), or those two libcalls will be expanded into scalar code. You can see the difference in [this commit](https://github.com/llvm/llvm-project/pull/99752/commits/f90a6b46a8a9639bad8f2aa826150a47727d6b82).
This is primarily for 32-bit targets calling i64 CTPOP.